### DISCUSSION TITLE
I am trying to get emcc to compile a cpp file which works fine under blazor...

### DISCUSSION THREAD

**Discussion by BradChase2011** [unoplatform/Uno.Wasm.Bootstrap#864] [Q&A]
I am seeking some help here to try to get over a hurdle using Uno.Wasm.Bootstrap instead of blazor to test out debugging and threading.  That said, I have this repository which works fine and compiles the cpp and js into the correct library:

https://github.com/BradChase2011/CSharpBlazorWebGPUExample

So what Ive done is switch the flags and such from these which work under blazor:

```
<EmccExtraLDFlags>-s USE_WEBGPU=1 -s NO_EXIT_RUNTIME=1 --js-library '.\lib\lib_webgpu.js'</EmccExtraLDFlags>
<NativeFileReference Include="lib\lib_webgpu_cpp20.cpp" />
```

and one of the calls is used like so:

`int test = navigator_gpu_get_preferred_canvas_format();`

into:

```
[DllImport("*")]
public static extern int navigator_gpu_get_preferred_canvas_format();
```

I have it compiling under Uno.Wasm.Bootstrap with these flags:

```
<WasmShellExtraEmccFlags Include="-s USE_WEBGPU=1 -s NO_EXIT_RUNTIME=1 --js-library '$(ProjectDir)\lib\lib_webgpu.js'" />
<WasmShellNativeCompile Include=".\lib\lib_webgpu_cpp20.cpp" />
```

Problem is when I go to call the imported function at run time it fails.  Under WSL its looking for an .so file by the same name (which doesnt exist anywhere) and running it outside of WSL just gives a generic error of failure.  So I guess my question is, how can I get that functionality that works under the blazor example app to run under uno wasm bootstrap, am I missing a setting or is this not possible?  If not possible, what would be the correct way?

Thanks much!

**Comment by jeromelaban** [comment]
Thanks for the report! Could you share your uno example as well?

**Reply by BradChase2011** [reply]
> Thanks for the report! Could you share your uno example as well?

I went ahead and put it up on github for ya:

https://github.com/BradChase2011/CSharpUnoWasmBootstrapWebGPUExample

I have the main work commented out in the program.cs file. For this test, it calls a simple function "navigator_gpu_get_preferred_canvas_format" that should return an int and fails.  If you see the error there, then the larger task would be uncommenting the main section in program.cs and seeing if it fails on blittable types (my next problem to look at).

Side note, I am not sure if its how I setup projects or my installation but I cannot debug the C# code using UNO or Blazor outside of WSL.  So if debug isnt working then switch over to WSL and debug there.  Also the errors will be different, same spot just different responses.

**Reply by jeromelaban** [reply]
Ok, so there are multiple things happening here. First, the method named `navigator_gpu_get_preferred_canvas_format` in C++ is located in  `lib_webgpu_dawn.cpp`, which is not included in your compilation, adding it in `WasmShellNativeCompile` will fail in current bootstrapper builds, but https://github.com/unoplatform/Uno.Wasm.Bootstrap/pull/865 will fix that, though this is not needed because of the javascript mention below.

Second, the bootstrapper does not support `DllImport("*")`, and you'll need to use `DllImport("__Native")` and add `<WasmShellAdditionalPInvokeLibrary Include="__Native" />` in the csproj.

Third, the method `navigator_gpu_get_preferred_canvas_format` is actually overriden by the export of the same name in `lib_webgpu.js`, which makes it call into javascript code rather than in C++ code, [here](https://github.com/BradChase2011/CSharpUnoWasmBootstrapWebGPUExample/blob/3277cf177ea348114cd976579ca1031e1ca76e73/lib/lib_webgpu.js#L604).

It's likely expected that the whole thing works this way, because most of what's in `dawn` needs to invoke javascript methods to work properly.

I'm able to invoke that method properly, so this should be working for you further along (regardless of https://github.com/unoplatform/Uno.Wasm.Bootstrap/pull/865 since you have only one cpp file to compile).

**Reply by BradChase2011** [reply]
Correct, ya it runs under blazor using that setup, so I tried to move the ideas over to the uno bootstrap.  I am not invoking the dawn code on either of them if I can recall correctly, which is why I am confused why it runs on the blazor side but not the uno.

That said, Im a little confused on the working part, are you saying you can call that import navigator_gpu_get_preferred_canvas_format from C#?  I took the comments out of the <WasmShellAdditionalPInvokeLibrary Include="__Native" /> in the csproj and changed the import to "[DllImport("__Native")]" but I get a new error when trying to run it at the navigator_gpu_get_preferred_canvas_format call:

System.DllNotFoundException: 'Unable to load shared library '__Native' or one of its dependencies. In order to help diagnose loading problems, consider using a tool like strace. If you're using glibc, consider setting the LD_DEBUG environment variable: 
/mnt/c/Development/ChaseSystems/CSharpUnoWasmBootstrapWebGPUExample/bin/Debug/net8.0/__Native.so: cannot open shared object file: No such file or directory

Or are you saying that the #865 fix is what allows using __Native and another compile "lib_webgpu_cpp20.cpp" in this case, and thats what is able to get you to hit the method?

Thanks again really appreciate it.

**Reply by BradChase2011** [reply]
Ohh weird... ok so in the project file I turned off WasmShellEnableThreads and the call worked for navigator_gpu_get_preferred_canvas_format using <WasmShellAdditionalPInvokeLibrary Include="__Native" /> and [DllImport("__Native")] but ONLY under the regular console dev web server.  Under WSL it still fails with:

System.DllNotFoundException: 'Unable to load shared library '__Native' or one of its dependencies. In order to help diagnose loading problems, consider using a tool like strace. If you're using glibc, consider setting the LD_DEBUG environment variable: 
/mnt/c/Development/ChaseSystems/CSharpUnoWasmBootstrapWebGPUExample/bin/Debug/net8.0/__Native.so: cannot open shared object file: No such file or directory

So to recap, I can hit and get the correct color format back using debug with the WasmShellAdditionalPInvokeLibrary of __Native, and WasmShellEnablethreads FALSE and the import set to __Native, as well as console dev web site.  It does NOT work under WSL and/or with threading turned on.


So what does that mean at this point?  Is there a way to be able to set multithreading on?  Run under WSL?  Keep debugging working?

Thanks!

EDIT:  I updated the example where it can call that first function but it still has the issue of running under WSL or with WasmShellEnableThreads enabled.
EDIT2:  I removed the non blittable bits which doesnt make sense as the struct was absolutely blittable.  I changed them to IntPtr so you can see the full issue when running, this will actually go through all the functions which fail in one way or another.

**Reply by jeromelaban** [reply]
If you're getting this particular error on WSL, it feels like this is not running the browser for some reason, but directly on linux. The browser should fail with a different error. Is this `System.DllNotFoundException` (particularly the `LD_DEBUG ` part) showing up in the browser logs?

**Reply by BradChase2011** [reply]
It is showing up in the debug output window when it stops on the C# line of the call.  Im new to the web stuff, is there somewhere else to check for the logs?  Does my explaniation of the threading make sense?  Is it something already known or do you have an idea why that could be stopping the calls?  The only reason I use WSL here and there is because I get better debugging support when multi thread is turned on.

edit:  BTW it appears the WSL is tacking on a linux assembly .so to it, "__Native.so"...  Should that be happening?

**Reply by jeromelaban** [reply]
I see. So debugging is not supported at all for debugging at this time (it will be improved with when the bootstrapper moves to net9). As for WSL, if it shows in the debuger window, it's likely that it's running on your linux directly and not in the browser. The impression that it gives it that debugger is better is because it uses the full .NET debugger, not the one specific to WebAssembly.

Which error do you get when running on windows, in the browser logs specifically? Try running without the debugger, it may give more information.

**Reply by BradChase2011** [reply]
Yea at this point ive been trying a ton of things to get around it.  The example app that I setup under .net 9 and blazor runs fine.  So the idea was to try and use the uno bootloader to try and see if threading or debugging was any better.  If you want to try and update the Uno example repository I put up and run that, it should fail quickly.

I get now with the changes you recommended on the example down to where it calls navigator_gpu_request_adapter_async, which fails.  Basically anything that uses an unmanaged delegate for a callback fails, which I moved the blazor example back to .net 8 and it actually makes it through that section after changing to IntPtrs.

We have a 3d engine and a multithreaded xaml engine(DirectX 12 and Vulkan) that we are trying to port to WebGPU.  Basically we are trying to prove each step, can we use C#, .NET, WASM, WebGPU, threading, and debugging.  We have the engine side written but cannot test due to failures at each stage of the tests.  Under blazor .net 9 we got the farthest but Microsoft has stated multithreading wont work until the earliest now .NET 10 and their debugging is completely broken at this point after the last update.  I guess thats the explanation of why we would love to get the Uno.Wasm.Bootloader working for us.  So the furthest we have been able to get is to the StartCalls() function of the Program, we are able to get the simple int returned from navigator_gpu_get_preferred_canvas_format but then it fails on the next call which uses an unmanaged delegate.  If we are able to get through that test in the example app, the next stage would be to test threading.  

If we turn on threading using WSL we actually hit the break points but fail because it cant find the library with the previous error.  If we dont use WSL with threading turned on, the debugger breaks.  So just trying to take it a step at a time and will worry about threading once we get the unmanaged callbacks working.  Currently when we call a function with an unmanaged callback like "navigator_gpu_request_adapter_async" it fails with this error:

Uncaught ManagedError 
    at marshal_exception_to_js (d:\a\Uno.DotnetRuntime.WebAssembly\Uno.DotnetRuntime.WebAssembly\runtime\src\mono\wasm\runtime\marshal-to-js.ts:349:18)
    at invoke_method_and_handle_exception (d:\a\Uno.DotnetRuntime.WebAssembly\Uno.DotnetRuntime.WebAssembly\runtime\src\mono\wasm\runtime\invoke-cs.ts:277:19)
    at runtimeHelpers.javaScriptExports.call_entry_point (d:\a\Uno.DotnetRuntime.WebAssembly\Uno.DotnetRuntime.WebAssembly\runtime\src\mono\wasm\runtime\managed-exports.ts:60:13)
    at mono_run_main (d:\a\Uno.DotnetRuntime.WebAssembly\Uno.DotnetRuntime.WebAssembly\runtime\src\mono\wasm\runtime\run.ts:44:45)
    at mainInit (localhost꞉50193/package_446f0aba010eaaef2687fe60cb3490713063d363/uno-bootstrap.js:363:30)


Forgive me as again I am new to the web stuff, so how should I check the browser logs?  I tried running without any debugger attached and chrome just shows the red background of the example, it doesnt show anything immediately seen as an error.  I setup a simple function with an unmanaged call back called test_case and the chrome developer tools show this only:

![image](https://github.com/unoplatform/Uno.Wasm.Bootstrap/assets/13279060/65f2c70d-38f3-46a1-8d41-c56e7679b48a)

I really do appreciate the help, I wish I could do more.

EDIT:  I wanted to add the simple JS call (test_Case) I made that uses a callback that works under blazor but not Uno that i referenced:

JS:
test_case__deps: [],
    test_case: function(callback) {
        {{{ makeDynCall('vii', 'callback') }}}(1, 2);
    return 3;
},

CS:
        [DllImport(_DllPath)]
        public static extern int test_case(delegate* unmanaged<int, int, void> callback);

int test = test_case(&TestCASE);

        [UnmanagedCallersOnly]
        static unsafe void TestCASE(int result, int userData)
        {
            int res = result;
        }

**Reply by jeromelaban** [reply]
I see. At this time, function pointers are not supported by p/invoke, this message is emitted silently here: 

```
[00:00:01.2281004] [] Skipping pinvoke 'WebGpu.Interop::navigator_gpu_request_adapter_async' because 'Parsing function pointer types in signatures is not supported.'.
```

This may be supported by .NET 9, and the bootstrapper is not using it yet.

Stil, it's possible work around it, even if it's not pretty, it allows for callback handlers to be called:

You need to change the interop for this:
```csharp
public unsafe struct WGpuRequestAdapterCallbackWrapper
{
    public WGpuRequestAdapterCallback cb;
}

public delegate void WGpuRequestAdapterCallback(WGpuAdapter* adapter, int userData);

[DlImport("__Native")]
/* EM_BOOL */
public static extern int navigator_gpu_request_adapter_async(ref WGpuRequestAdapterOptions options, WGpuRequestAdapterCallbackWrapper adapterCallback, IntPtr userData);
```

Then add this attribute, somewhere in your assemb;y:

```csharp
[AttributeUsage(AttributeTargets.Method)]
internal sealed class MonoPInvokeCallbackAttribute : Attribute
{
    public MonoPInvokeCallbackAttribute(Type type)
    {
        Type = type;
    }

    public Type Type { get; private set; }
}
```

then modify your call back to be:

```csharp
[MonoPInvokeCallback(typeof(WGpuRequestAdapterCallback))]
static unsafe void ObtainedWebGpuAdapter(WGpuAdapter* result, int userData)
```

then modify your invocation to be:

```csharp
navigator_gpu_request_adapter_async(ref options, new() { cb = ObtainedWebGpuAdapter }, IntPtr.Zero);
```

Note that I changed `ObtainedWebGpuAdapter` to take a pointer, since `WGpuAdapter` is a pointer itself:

```cpp
typedef int WGpuObjectBase;
```

All of this is required for now because `UnmanagedCallers` has semantics that requires function pointers, and .NET 8 does not support it properly.

**Reply by BradChase2011** [reply]
Thanks much for the update!  I had no idea that was only supported in .net 9, which explains why the blazor side works with the latest pre releases but not the uno side which is .net 8.  Man wish I had known that, ok so Ill try to implement this stuff and rewrite all the callbacks after a test.  Wish I had seen that error *face palm* haha.

**Reply by BradChase2011** [reply]
All the callbacks work great so far!  I have another issue that has come up though dealing with Structs.  In the blazor example, I can send back a struct marshalled correctly but when I go to return the struct from the uno side with .net 8 I can only return a one field size struct, anything larger It fails.  I presume this has to deal with some sort of marshalling or maybe the location of the memory?  Here is a quick setup of a test struct I made.  Im sure I am missing something again that exists in .net 9 blazor that isnt available in .net 8, Anything up front you see as not acceptable?

Here is the example:

cpp:

`typedef struct TestStruct
{
    int first;
    int second;
} TestStruct;

TestStruct GetTestStruct()
{
    TestStruct ret;
    ret.first = 1;
    ret.second = 2;
    return ret;
}`


C# Interop:

`        [StructLayout(LayoutKind.Sequential)]
        public struct TestStruct
        {
            public int first;
            public int second;
        }


        [DllImport(_DllPath)]
        public static extern TestStruct GetTestStruct();`


C# Call:

`TestStruct testStruct = GetTestStruct();`


I have tried setting the size manually, explicitly, types, I cant seem to get it to pass through.  The final error I get is "Uncaught RuntimeError RuntimeError: null function or function signature mismatch".  This seems to be the last hold out as of yet so the webgpu side can provide the defaults back to the client.  Thanks again, I really appreciate all the help you have provided.

EDIT:  I forgot to add, I can pass them as a ref as input to the function and that works just fine setting it there in the cpp side but I would like to know if maybe there is something small I missed to be able to actually return the struct rather than setting the memory from reference.  Maybe something like a mono flag or attribute?  For reference the UNO example is checked in and working with the latest changes showing webgpu running.
